### PR TITLE
Replace inline new version dialog with upload flow a-la new addon

### LINKED ISSUE #3469: New version update submission flow for Listed and Unlisted
Part of #3465

The add-on version update submission flow will have a look and feel more consistent with the new add-on submission flow. Also, at this point it will be possible to choose a different channel for new versions, which means `channel` will no longer be set based on `is_listed`. UX for this new flow is in mozilla/addons#192.

### PR DESCRIPTION
**Pull request by eviljeff**
Fixes #3469

### CODE REVIEW

File: src/olympia/devhub/tests/test_views_submit.py [hunk 1]
```diff
@@ -110,11 +113,45 @@ def test_read_dev_agreement_skip(self):
         self.assert3xx(response, reverse('devhub.submit.distribution'))
 
 
-class TestCreateAddon(BaseUploadTest, TestCase):
+class TestAddonSubmitDistribution(TestCase):
     fixtures = ['base/users']
 
     def setUp(self):
-        super(TestCreateAddon, self).setUp()
+        super(TestAddonSubmitDistribution, self).setUp()
```
Comment (eviljeff) [inline src/olympia/devhub/tests/test_views_submit.py]: This isn't a new class - it was just in the wrong order (you choose the distribution before you create the Addon).  See below in the file.

File: src/olympia/devhub/tests/test_views_submit.py [hunk 2]
```diff
@@ -763,3 +767,444 @@ def test_redirect_from_other_pages(self):
         r = self.client.get(reverse('devhub.addons.edit', args=['a3615']),
                             follow=True)
         self.assert3xx(r, reverse('devhub.submit.details', args=['a3615']))
+
```
Comment (eviljeff) [inline src/olympia/devhub/tests/test_views_submit.py]: new tests from here on.  The functionality that's been tested is similar to the tests for the old flow in test_views.py though and we'll be able to delete those once that code is deleted.

File: src/olympia/devhub/urls.py [hunk 3]
```diff
@@ -11,15 +11,6 @@
 PACKAGE_NAME = '(?P<package_name>[_\w]+)'
 
 
-# These will all start with /addon/<addon_id>/submit/
```
Comment (eviljeff) [inline src/olympia/devhub/urls.py]: flattened this because it didn't seem worth it for only 2 sub patterns.

File: src/olympia/devhub/views.py [hunk 4]
```diff
@@ -1387,109 +1390,207 @@ def version_stats(request, addon_id, addon):
...
+@login_required
+def submit_addon_upload(request, channel):
+    channel_id = amo.CHANNEL_CHOICES_LOOKUP[channel]
+    return _submit_upload(request, None, channel_id,
+                          'devhub.submit.details', 'devhub.submit.finish')
+
+
+@dev_required
+@waffle_switch('step-version-upload')
+def submit_version_upload(request, addon_id, addon, channel):
```
Comment (eviljeff) [inline src/olympia/devhub/views.py]: really the only difference (now) between this as `submit_version` is the url pattern needs a channel param:
`^versions/submit/upload-(?P<channel>listed|unlisted)$` vs. `^versions/submit/$`.  I'm undecided on ideal urls.

File: src/olympia/versions/models.py [hunk 5]
```diff
@@ -145,7 +145,7 @@ def from_upload(cls, upload, addon, platforms, channel, send_signal=True,
         )
         log.info(
             'New version: %r (%s) from %r' % (version, version.id, upload))
-
+        amo.log(amo.LOG.ADD_VERSION, version, addon)
```
Comment (eviljeff) [inline src/olympia/versions/models.py]: we've not logged version creation since June 2013!

Comment (diox) [inline src/olympia/versions/models.py]: /o\

File: src/olympia/addons/forms.py [hunk 6]
```diff
@@ -296,9 +296,9 @@ def __init__(self, *args, **kw):
             if not acl.action_allowed(self.request, 'Addons', 'Edit'):
                 form.disabled = (app and self.addon.is_featured(app))
 
-    def save(self):
+    def save(self, addon=None):
```
Comment (diox) [inline src/olympia/addons/forms.py]: I'm confused. Why would you need to pass `addon` to the `save()` method when this form already takes an `addon` at init ?

File: src/olympia/devhub/templates/devhub/addons/submit/base.html [hunk 7]
```diff
@@ -1,7 +1,10 @@
 {% extends "devhub/base.html" %}
 
-
-{% set title = _('Submit a New Add-on') %}
+{% if submit == 'version' %}
```
Comment (diox) [inline src/olympia/devhub/templates/devhub/addons/submit/base.html]: Not a big fan of calling it just `submit`. Maybe something like `submit_page` ?

File: src/olympia/devhub/templates/devhub/addons/submit/upload.html [hunk 8]
```diff
@@ -38,7 +56,7 @@
         {{ new_addon_form.errors.supported_platforms }}
         {{ new_addon_form.supported_platforms }}
       </div>
-      {{ new_addon_form.upload }}
```
Comment (diox) [inline src/olympia/devhub/templates/devhub/addons/submit/upload.html]: I looked at git log and tried various scenarios locally and it doesn't seem to be needed at all, my guess is that it was just not removed when the upload flow last changed... 5 years ago.

File: src/olympia/devhub/urls.py (hunk 3 above)
Comment (diox) [inline src/olympia/devhub/urls.py]: :+1:

File: src/olympia/devhub/views.py [hunk 9]
```diff
@@ -1387,109 +1390,207 @@ def version_stats(request, addon_id, addon):
...
+        addon=addon,
         request=request
     )
-    is_listed = channel == 'listed'
-    if request.method == 'POST':
-        if form.is_valid():
-            data = form.cleaned_data
-
-            p = data.get('supported_platforms', [])
+    if request.method == 'POST'and form.is_valid():
```
Comment (diox) [inline src/olympia/devhub/views.py]: Missing space between `'POST'` and `and`. Surprised `flake8` didn't catch this.

File: src/olympia/devhub/forms.py [hunk 10]
```diff
@@ -557,23 +540,31 @@ class StandaloneValidationForm(AddonUploadForm):
...
+
     beta = forms.BooleanField(
         required=False,
         help_text=_lazy(u'A file with a version ending with '
                         u'a|alpha|b|beta|pre|rc and an optional number is '
                         u'detected as beta.'))
 
     def __init__(self, *args, **kw):
-        self.addon = kw.pop('addon')
+        self.addon = kw.pop('addon', None)
```
Comment (diox) [inline src/olympia/devhub/forms.py]: Why ? When should this happen ? This seem like a bad idea.

Comment (eviljeff) [inline src/olympia/devhub/forms.py]: `NewVersionForm` is used (now) both for a new version on an existing Addon and for a new Version on a new Addon (e.g. a new Addon).  In the latter case `addon` doesn't exist at this point.  Could pass in `addon=None` in the constructor but it seems less clear what's going on.

File: src/olympia/devhub/views.py [hunk 11]
```diff
@@ -1387,109 +1390,207 @@ def version_stats(request, addon_id, addon):
...
+                source=data['source'],
+                is_beta=is_beta)
+            url_args = [addon.slug, version.id]
+        else:
+            addon = Addon.from_upload(
+                upload=data['upload'],
+                platforms=data.get('supported_platforms', []),
+                source=data['source'],
+                is_listed=channel == amo.RELEASE_CHANNEL_LISTED)
+            version = addon.current_version
```
Comment (diox) [inline src/olympia/devhub/views.py]: Use `.find_latest_version(channel=channel)` instead

File: src/olympia/devhub/views.py [hunk 12]
```diff
@@ -1387,109 +1390,207 @@ def version_stats(request, addon_id, addon):
...
-    is_admin = acl.action_allowed(request, 'ReviewerAdminTools', 'View')
+            url_args = [addon.slug]
 
+        check_validation_override(request, form, addon, version)
+        if data['source']:
+            addon_update = {'admin_review': True}
+        else:
+            addon_update = {}
+        if addon.status == amo.STATUS_NULL and addon.has_complete_metadata():
+            addon_update.update(status=amo.STATUS_NOMINATED)
```
Comment (diox) [inline src/olympia/devhub/views.py]: You can just use `addon_update['status'] = amo.STATUS_NOMINATED`

File: src/olympia/devhub/views.py [hunk 13]
```diff
@@ -1387,109 +1390,207 @@ def version_stats(request, addon_id, addon):
...
-                return redirect('devhub.submit.finish', addon.slug)
-            return redirect('devhub.submit.details', addon.slug)
-    is_admin = acl.action_allowed(request, 'ReviewerAdminTools', 'View')
+            url_args = [addon.slug]
 
+        check_validation_override(request, form, addon, version)
+        if data['source']:
+            addon_update = {'admin_review': True}
+        else:
+            addon_update = {}
```
Comment (diox) [inline src/olympia/devhub/views.py]: Might be slightly easier to read and more future-proof to always do `addon_update = {}` above, then add things to the dict if necessary. Something like:
```python
addon_update = {}
if data['source']:
    addon_update['admin_review'] = True
if addon.status == amo.STATUS_NULL and addon.has_complete_metadata():
    addon_update['status'] = amo.STATUS_NOMINATED

```

File: src/olympia/devhub/views.py (hunk 11 above)
Comment (eviljeff) [inline src/olympia/devhub/views.py]: whoops

File: src/olympia/devhub/views.py [hunk 14]
```diff
@@ -1387,109 +1390,207 @@ def version_stats(request, addon_id, addon):
...
+        context.update(license_form.get_context())
+        context.update(form=describe_form, cat_form=cat_form)
+        forms_list.extend([describe_form, cat_form, context['license_form']])
+    reviewer_form = forms.VersionForm(
         post_data, instance=latest_version)
+    context.update(reviewer_form=reviewer_form)
+    forms_list.extend([reviewer_form])
+
+    if request.method == 'POST' and all(
+            [form.is_valid() for form in forms_list]):
```
Comment (diox) [inline src/olympia/devhub/views.py]: You shouldn't need the `[]`, you can just do `all(form.is_valid() for form in forms_list)`

File: src/olympia/devhub/views.py [hunk 15]
```diff
@@ -1387,109 +1390,207 @@ def version_stats(request, addon_id, addon):
...
+            post_data, addon=addon, request=request)
+        license_form = forms.LicenseForm(
+            post_data, version=latest_version, prefix='license')
+        context.update(license_form.get_context())
+        context.update(form=describe_form, cat_form=cat_form)
+        forms_list.extend([describe_form, cat_form, context['license_form']])
+    reviewer_form = forms.VersionForm(
         post_data, instance=latest_version)
+    context.update(reviewer_form=reviewer_form)
+    forms_list.extend([reviewer_form])
```
Comment (diox) [inline src/olympia/devhub/views.py]: `forms_list.append(reviewer_form)`

File: src/olympia/devhub/forms.py (hunk 10 above)
Comment (diox) [inline src/olympia/devhub/forms.py]: I thought so but couldn't find where it's instantiated without an addon ?

Comment (eviljeff) [inline src/olympia/devhub/forms.py]: Ah... in `_submit_upload` the `addon` parameter is `None` for new Addon uploads, and that's passed onto `forms.NewVersionForm`.  So I can either change that to not pass the `addon` parameter if `None` or take out the `self.addon = kw.pop('addon', None)` in the form `__init__`

Comment (diox) [inline src/olympia/devhub/forms.py]: Not a super big deal now that the `save() method doesn't accept an `addon` parameter. I think I'd revert the `kw.pop('addon', None)` change but it's up to you.

Comment (eviljeff) [inline src/olympia/devhub/forms.py]: The save() method (not) accepting an `addon` parameter is a different form (BaseCategoryFormSet)

**Pull request merged**